src/core/tranche/distributor/maker.sol

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2019 Centrifuge
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.4.24;
+
+import "ds-note/note.sol";
+import "../distributor.sol";
+import "../../lightswitch.sol";
+
+contract MakerDistributor is Distributor, Switchable {
+
+    function balance() public auth {
+
+//        require(flowThrough==false);
+
+        if (manager.poolClosing()) {
+            repayTranches();
+        } else {
+            poolOpen();
+        }
+    }
+
+    function poolOpen() private {
+        int wad = manager.checkPile();
+        if (wad > 0) {
+            borrowFromTranches(uint(wad));
+        } else if (wad < 0) {
+            repayTranches();
+        }
+    }
+
+    function borrowFromTranches(uint wad) private {
+        uint borrowAmount = wad;
+        for (uint i = 0; i < manager.trancheCount(); i++) {
+            OperatorLike o = OperatorLike(manager.operatorOf(i));
+            uint trancheReserve = o.balance();
+            if (borrowAmount < trancheReserve) {
+                o.borrow(address(manager.pile), borrowAmount);
+                return;
+            }
+            o.borrow(address(manager.pile), trancheReserve);
+            borrowAmount = sub(borrowAmount, trancheReserve);
+        }
+    }
+
+    // Note: assumes two tranche setup for now
+
+    // max_take is how much liquidity can be taken out from a specific tranche, given the current equity reserve/equity debt,
+    // in order to maintain the equity ratio which has been set by the pool manager.
+
+    // max_take =  (Equity.Reserve + Equity.Debt)/Equity.Ratio * Senior.Ratio - Senior.Debt
+
+    function calcMaxTake() public auth returns (uint) {
+        OperatorLike oE = OperatorLike(manager.operatorOf(0));
+        OperatorLike oS = OperatorLike(manager.operatorOf(1));
+
+        uint wadER = oE.balance();
+        uint wadED =  oE.debt();
```

Comment:
I think this one is wrong. We need the `real` debt not the expected `debt` from equity tranche.
`oE.debt()` is not covering the losses.